### ISSUE TITLE
First fast then slow down dramatically

### ISSUE THREAD

**Issue by ZHUANGHP** [NVIDIA/DALI#1325]
I installed the DALI and test it with resnet18/resnet34 on ImageNet using the pytorch [resnet example](https://github.com/NVIDIA/DALI/blob/master/docs/examples/pytorch/resnet50/main.py). The experiment went quite well at first (up to 700+ images/s for resenet18 with batch size 256) but after a few mins, the speed drops to, say 120 images/s and stay that way (see the following example for resnet18). This phenomenon exists no matter how many workers I choose.

Epoch: [0][830/5004]    Time 0.380 (0.402)      **Speed** 674.385 (636.991) Data 0.023 (0.036)      Loss 6.7732 (6.8701)    Prec@1 0.391 (0.369)    Prec@5 1.953 (1.490)
Epoch = 0, step = 840, lr = 0.003361310951239009
Epoch: [0][840/5004]    Time 0.399 (0.402)      **Speed** 642.141 (636.903) Data 0.024 (0.036)      Loss 6.7014 (6.8680)    Prec@1 0.000 (0.373)    Prec@5 2.344 (1.503)
Epoch = 0, step = 850, lr = 0.0034012789768185455
Epoch: [0][850/5004]    Time 0.407 (0.402)      **Speed** 629.342 (636.775) Data 0.030 (0.036)      Loss 6.7017 (6.8659)    Prec@1 1.562 (0.381)    Prec@5 3.125 (1.519)
Epoch = 0, step = 860, lr = 0.003441247002398082
Epoch: [0][860/5004]    Time 0.398 (0.402)      **Speed** 642.648 (636.699) Data 0.035 (0.036)      Loss 6.6860 (6.8638)    Prec@1 0.781 (0.387)    Prec@5 3.125 (1.542)
Epoch = 0, step = 870, lr = 0.0034812150279776182
Epoch: [0][870/5004]    Time 1.793 (0.416)      **Speed** 142.758 (615.546) Data 1.442 (0.050)      Loss 6.6027 (6.8615)    Prec@1 1.953 (0.396)    Prec@5 5.469 (1.560)
Epoch = 0, step = 880, lr = 0.0035211830535571546
Epoch: [0][880/5004]    Time 1.181 (0.427)      **Speed** 216.794 (599.244) Data 0.832 (0.062)      Loss 6.6737 (6.8593)    Prec@1 0.000 (0.403)    Prec@5 3.906 (1.580)
Epoch = 0, step = 890, lr = 0.003561151079136691
Epoch: [0][890/5004]    Time 1.343 (0.438)      **Speed** 190.604 (584.027) Data 0.995 (0.073)      Loss 6.6696 (6.8572)    Prec@1 0.781 (0.407)    Prec@5 2.734 (1.602)
The disk containing the dataset is an HDD. Interestingly, I move the ImageNET to an SSD and the speed can be maintained. 

anyone know the reason? Thank you

**Comment by JanuszL** [comment]
Hi,
Can you check your CPU/GPU utilization and disc IO stats?
I guess that the first 860 images are cached in your OS disc cache while the rest need to be read from the drive. If you don't have fast SSD then it is the bottleneck, not the processing itself.

**Comment by ZHUANGHP** [comment]
@Januszl the CPU/GPU utilizations are high when the reading speed is fast. haven't checked the io.The GPU util is near 0 when the slow-down happens.

One more thing, the SSD case also faces slow-down (900+ to 600+) after about 10mins but it is probably because of the PC generating too much heat.

**Comment by JanuszL** [comment]
So I'm almost sure your are hitting disc IO bottleneck.

**Comment by ZHUANGHP** [comment]
@JanuszL It might or might not be the case actually. You pointed out that the OS might cache 860 images in advance, but it is not 860 images, it should be 860*256 images. Now I am a bit surprised that the OS can cache this amount of data very instantly (building the dataloader costs approx. 2s).

**Comment by JanuszL** [comment]
Right 860*256. It can keep it cached from the previous runs. Again please check disc IO first and you can read more about it [here](https://blog.webnersolutions.com/understanding-disk-cache-in-linux).

**Comment by ZHUANGHP** [comment]
Ok. Will let you know the details.

**Comment by ZHUANGHP** [comment]
@JanuszL update:
I checked the IOs. It turns out that the reading is rather slow (around 100M/s) with an SKhynix SC300 SSD. Also, I manually equip the machine with 16G of extra memory. The training is for now stabilized.

To summarize, I think the culprits are the low-performing SSD and the lack of memory to accommodate DALI.

Will let you know other updates.

**Comment by JanuszL** [comment]
> To summarize, I think the culprits are the low-performing SSD and the lack of memory to accommodate DALI.

If you don't have enough memory for DALI then you just get OOM. In your case, it was the thing independent from DALI. If not DALI then CPU based processing probably would have hide low reading speed.
